### PR TITLE
build: compile wasm udfs automatically

### PR DESCRIPTION
**Pull request by wmitros**
Currently, when we deal with a Wasm program, we store
it in its final WebAssembly Text form. This causes a lot
of code bloat and is hard to read. Instead, we would like
to store only the source codes, and build Wasm when
necessary. This series adds build commands that
compile C/Rust sources to Wasm and uses them for Wasm
programs that we're already using.

After these changes, adding a new program that should be
compiled to Rust, requires only adding the source code
of it and updating the `wasms` and `wasm_deps` lists in
`configure.py`.

All Wasm programs are build by default when building all
artifacts, artifacts in a given mode, or when building
tests. Additionally, a {mode}-wasm target is added, so that
it's possible to build just the wasm files.
The generated files are saved in $builddir/{mode}/wasm,
and are accessed in cql-pytests similarly to the way we're
accessing the scylla binary - using glob.

### CODE REVIEW

**Comment by wmitros** [comment]
Before this change, we should update the toolchain with https://github.com/scylladb/scylladb/pull/12896

File: configure.py [hunk 1]
```diff
@@ -1846,6 +1850,16 @@ def query_seastar_flags(pc_file, link_static_cxx=False):
               command = CARGO_BUILD_DEP_INFO_BASEDIR='.' cargo build --locked --manifest-path=rust/Cargo.toml --target-dir=$builddir/{mode} --profile=rust-{mode} $
                         && touch $out
               description = RUST_LIB $out
+            rule c2wasm
+                command = clang --target=wasm32 --no-standard-libraries -Wl,--export-all -Wl,--no-entry $in -o $out
+                description = C2WASM $out
+            rule rust2wasm
+                command = RUSTFLAGS="-C link-args=-zstack-size=131072" cargo build --target=wasm32-wasi --example=$example --locked --manifest-path=test/resource/wasm/rust/Cargo.toml --target-dir=$builddir/{mode}/wasm/ $
```
[the remark this answers is not in the document]
Comment (avikivity) [reply configure.py:1875]: Might as well add the comment to the source.

File: configure.py [hunk 2]
```diff
@@ -1869,6 +1883,16 @@ def query_seastar_flags(pc_file, link_static_cxx=False):
...
                 continue
+            if binary in wasms:
+                src = wasm_deps[binary]
+                wasm = binary[:-4] + '.wasm'
+                if src.endswith('.rs'):
+                    f.write('build $builddir/{}/{}: rust2wasm.{} {} | test/resource/wasm/rust/Cargo.lock\n'.format(mode, wasm, mode, src))
+                    f.write('   example = {}\n'.format(binary[binary.rindex('/')+1:-4]))
+                else:
+                    f.write('build $builddir/{}/{}: c2wasm.{} {}\n'.format(mode, wasm, mode, src))
+                f.write('build $builddir/{}/{}: wasm2wat.{} $builddir/{}/{}\n'.format(mode, binary, mode, mode, wasm))
```
Comment (avikivity) [inline configure.py:1912]: New code should use f-strings, not format().

File: configure.py [hunk 3]
```diff
@@ -1951,11 +1975,18 @@ def query_seastar_flags(pc_file, link_static_cxx=False):
                 header_objs=' '.join(["$builddir/{mode}/{hh}.o".format(mode=mode, hh=hh) for hh in headers])
             )
         )
+        f.write(
+            'build {mode}-wasm: phony {wasms}\n'.format(
+                mode=mode,
+                wasms=' '.join(['$builddir/{}/{}'.format(mode, binary) for binary in sorted(wasms)]),
+            )
```
Comment (avikivity) [inline configure.py:2000]: I don't think wasms should be mode-specific, they don't have distinct flags, do they?

File: configure.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (wmitros) [reply configure.py:1875]: Makes sense

File: configure.py (hunk 2 above)
Comment (wmitros) [reply configure.py]: Ok

**Comment by wmitros** [comment]
In the rebase I fixed a conflict with master and added calls to `wasm-opt` and `wasm-strip` to make the `wasm` binaries smaller for a relatively small cost

**Comment by nyh** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/214/

```
/bin/sh: line 1: wasm-opt: command not found
```

**Comment by wmitros** [comment]
> > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/214/
> 
> ```
> /bin/sh: line 1: wasm-opt: command not found
> ```

Yes, I submitted https://github.com/scylladb/scylladb/pull/13282

**Comment by avikivity** [comment]
wasm should be mode-independent, our wasm compiles shouldn't be affected by dev/debug/release. Please address in a follow-up.

File: configure.py [hunk 4]
```diff
@@ -563,6 +563,9 @@ def find_headers(repodir, excluded_dirs):
     'test/raft/failure_detector_test',
 ])
 
+wasms = set([
+])
+
```
Comment (avikivity) [inline configure.py:576]: /cc @tchaikov. How does cmake built rust and wasm?

File: configure.py [hunk 5]
```diff
@@ -1350,6 +1358,16 @@ def find_headers(repodir, excluded_dirs):
...
 
+wasm_deps['wasm/return_input.wat'] = 'test/resource/wasm/rust/return_input.rs'
+wasm_deps['wasm/test_short_ints.wat'] = 'test/resource/wasm/rust/test_short_ints.rs'
+
+wasm_deps['wasm/test_fib_called_on_null.wat'] = 'test/resource/wasm/c/test_fib_called_on_null.c'
+wasm_deps['wasm/test_mem_grow.wat'] = 'test/resource/wasm/c/test_mem_grow.c'
+wasm_deps['wasm/test_pow.wat'] = 'test/resource/wasm/c/test_pow.c'
+wasm_deps['wasm/test_UDA_final.wat'] = 'test/resource/wasm/c/test_UDA_final.c'
+wasm_deps['wasm/test_UDA_scalar.wat'] = 'test/resource/wasm/c/test_UDA_scalar.c'
+wasm_deps['wasm/test_word_double.wat'] = 'test/resource/wasm/c/test_word_double.c'
```
Comment (avikivity) [inline configure.py:1369]: Idea for follow-up: name the targets `wasm/return_input.rs.wat`, and then you can infer the main dependency from the target name and can drop these.

**Comment by wmitros** [comment]
> wasm should be mode-independent, our wasm compiles shouldn't be affected by dev/debug/release. Please address in a follow-up.

I actually managed to accomplish this in this patch already, as mentioned in https://github.com/scylladb/scylladb/pull/13209#discussion_r1139109173

**Comment by avikivity** [comment]
> > wasm should be mode-independent, our wasm compiles shouldn't be affected by dev/debug/release. Please address in a follow-up.
> 
> I actually managed to accomplish this in this patch already, as mentioned in [#13209 (comment)](https://github.com/scylladb/scylladb/pull/13209#discussion_r1139109173)

I'm so predictable.

**Pull request merged**